cloudify_cli/replace_certificates_config.py

```diff
@@ -0,0 +1,377 @@
+########
+# Copyright (c) 2020 Cloudify.co Ltd. All rights reserved
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+############
+
+import os
+import abc
+import time
+import select
+from contextlib import contextmanager
+
+import yaml
+
+from . import env
+from .constants import CERTS_MAPPING, NEW_CERTS_TMP_DIR_PATH
+from .exceptions import CloudifyCliError
+
+try:
+    from paramiko import SSHClient, AutoAddPolicy
+except ImportError:
+    SSHClient = None
+
+CONFIG_FILE_PATH = '{0}replace_certificates_config.yaml'
+LOCAL_CONFIG_FILE_PATH = '/tmp/' + CONFIG_FILE_PATH
+REMOTE_CONFIG_FILE_PATH = NEW_CERTS_TMP_DIR_PATH + CONFIG_FILE_PATH.format('')
+
+
+def retry_with_sleep(func, *func_args, **kwargs):
+    retry_count = kwargs.get('retry_count', 15)
+    delay = kwargs.get('delay', 2)
+    for i in range(retry_count):
+        try:
+            return func(*func_args)
+        except Exception as e:
+            if i < retry_count - 1:
+                time.sleep(delay)
+                continue
+            else:
+                raise e
+
+
+class Node(object):
+    def __init__(self,
+                 host_ip,
+                 username,
+                 key_file_path,
+                 new_cert_path,
+                 new_key_path,
+                 new_ca_cert_path,
+                 logger):
+        self.host_ip = host_ip
+        self.username = username
+        self.key_file_path = key_file_path
+        self.client = self._create_ssh_client()
+        self.new_cert_path = new_cert_path
+        self.new_key_path = new_key_path
+        self.new_ca_cert_path = new_ca_cert_path
+        self.certs_mapping = self._create_paths_mapping()
+        self.certs_dict = {}
+        self.errors_list = []
+        self.logger = logger
+
+    @staticmethod
+    def _create_ssh_client_func(hostname, username, key_file):
+        client = SSHClient()
+        client.set_missing_host_key_policy(AutoAddPolicy())
+        client.connect(hostname=hostname, username=username,
+                       key_filename=key_file)
+        return client
+
+    def _create_ssh_client(self):
+        return retry_with_sleep(self._create_ssh_client_func,
+                                self.host_ip, self.username,
+                                self.key_file_path)
+
+    def _blocking_exec_command(self, command):
```

Comment:
are you sure we want to use this, and not just reuse existing fabric stuff, this seems pretty low-level for what we want to do

Reply:
Changed back to fabric. The purpose of this function is to print the output as it is printed at the remote host, but with fabric it's automatic.